cpp/include/cudf/io/writers.hpp

```diff
@@ -222,7 +222,150 @@ class writer {
 
 }  // namespace parquet
 
+//! CSV format
+namespace csv {
 
+/**
+ * @brief Options for the CSV writer. 
+ * Also base class for `write_csv_args`
+ */
+struct writer_options {
```

Comment:
Perhaps I'm missing something, but why use a struct here instead of a class? Alternatively, why are the data members private? I don't see an invariant that needs to be preserved with setters.

Reply:
In C++ a `struct` is just a class where members are by default `public` (rather than `private`; and inheritance is by default `public`, etc.) This is not a POD (aggregate) C-style struct. What's wrong with a `struct`?

Data members _should_ always be private (encapsulation-wise, etc.) I believe I had a presentation on exactly this topic in the "Culture of Quality" series. This is not something I cooked. It's coming down from the likes of Scott Meyers, etc. ("Effective C++", Scott Meyers, 2nd ed. Item 20). You're opening the door for a lot of headaches down the line by making data members public and not accessing them via accessors.

Also, the accessors are inlined, so they've 0 overhead cost at runtime.

Reply:
I still don't see why you would use a struct here, instead of a class. I know of two criteria to pick class vs. struct. One is to check if there are invariant (that ctor+setters enforce). The other one is to use classes when inheritance is involved. Struct dose not fix either of these two criteria.

> You're opening the door for a lot of headaches down the line by making data members public and not accessing them via accessors.

I don't really see how, in this case, but if encapsulation is something we prefer in cuDF, I don't mind it that much. Just seems to add unnecessary code here IMO.

Reply:
I don't understand what your case against `struct` is. Whatever you can do with a `class` you can with a `struct` (inherit it, pass it as template argument, etc.).  

>  One is to check if there are invariant (that ctor+setters enforce). The other one is to use classes when inheritance is involved.

Can you please elaborate on the above?

Reply:
> I don't really see how, in this case, but if encapsulation is something we prefer in cuDF, I don't mind it that much. Just seems to add unnecessary code here IMO.

What if you currently access a data member directly, say, `obj.field` and sometimes in the future you need to replace that `field` with a lengthy computation? You'd have to replace all the occurrences (maybe thousands) of `obj.field` in your code with a function (or a block of code). While if you used an accessor `obj.field()` in the 1st place, you'd only have to change the definition of that accessor `field()`.

Reply:
Actually, I think for *_options and *_args, we'd want to use structs with everything public, since it's just a way to group the various parameters and are likely to be accessed from the various python/cython/java bindings, where IIRC being able to set individual fields is desirable.

Reply:
> Actually, I think for *_options and *_args, we'd want to use structs with everything public

I agree with this, I don't think there is a practical reason for encapsulation.

>What if you currently access a data member directly, say, obj.field and sometimes in the future you need to replace that field with a lengthy computation?

I'm aware of these concerns and absolutely agree with them in general. But I don't see such a scenario could happen with our _args/_options.

Reply:
On the subject of struct vs. class: As you said, there is barely any difference between struct and class. Because of this, people are using the available choice to convey meaning not included in the language:
1) structs are aggregates, classes have invariants/behavior (e.g. https://www.fluentcpp.com/2017/06/13/the-real-difference-between-struct-class/) -makes sense w.r.t. the default member visibility.
2) classes are used for inheritance - maybe makes sense w.r.t. C vs. C++, but IIUC the goal is to use the struct/class choice to help the reader see if a type is a part of a hierarchy.

There are probably more conventions I am not aware of, and that's why I asked why you used struct here. I'm pushing for any specific convention, but would prefer for us to pick and follow one.

Reply:
>     2\. classes are used for inheritance - maybe makes sense w.r.t. C vs. C++, but IIUC the goal is to use the struct/class choice to help the reader see if a type is a part of a hierarchy.

This is a matter of style. For me, `struct` is just a convenient way to avoid the `public` qualifier, which is the default one in `struct`. It never has the meaning of an aggregate, because aggregates (POD) have no place in C++ code, IMO. They're C leftovers. (I would) Avoid them. Also, semantically and, in fact, _syntactically_ speaking: "_Aggregates are either arrays or simple, C-like classes that have no user-provided constructors, no private or protected
non-static data members, no virtual functions, and before C++17, no base classes._"  (Josuttis, "C++17 The Complete Guide"). So, the simple fact that a `struct` has `private` members it's a clear indication it is _not_ an aggregate,

That being said, at some point pedantic debates must end and concrete decisions must be taken. To that end, I'm getting a bit confused about what's desired here: `struct` for `args` and `options` and `class` for everything else; and everything under the Sun should be public? Is that the desired change request? If not, please clarify.

Reply:
> For me, struct is just a convenient way to avoid the public qualifier, which is the default one in struct.

Got it. I don't agree, but I understand where this is coming from.

> and everything under the Sun should be public?

Not sure what this means. The suggestion is to have `_args` and `_options` be structs with public data members. I think it makes the usage and maintenance a bit easier, and makes this more consistent with other formats' args and options.

If you feel strongly against this, feel free to ignore my suggestion. I expect we will soon have a separate PR to clean up the args/options for all IO formats (apply the inheritance pattern you used here, etc.) and then we can decide on details like this, and make everything consistent. Within the scope of this PR, I'm okay with the current setup.

Reply:
Done. I made args/options data members public.